### ISSUE TITLE
Upgrade WebOB (and webapp2 to be safe)

### ISSUE THREAD

**Issue by nworden** [google/personfinder#575]
Our current Docker setup is running Python v2.7.12 (because that's what's in the Ubuntu Xenial repository). It breaks with Python 2.7.15 (I haven't tried versions in between).

The reason is that the newer version of `xml.sax.expatreader.ExpatParser` (used by `app/pfif.py`) expects to be able to call `close` on its input file, and the input object provided by WebOB v1.1.1 (a `LimitedLengthFile`) doesn't have a `close` function. It does have that function in WebOB v1.2.3, so we should upgrade to that. It might be a good idea to upgrade to webapp2 v2.5.2 at the same time (since they're closely linked it seems better to run versions from similar times).

**Comment by nworden** [comment]
Once I get things working with Python 2.7.15, I'll see if I can set up Travis to run with two environments: one that uses 2.7.12 (because that seems to be [what App Engine currently uses](https://cloud.google.com/appengine/docs/standard/python/release-notes) for its 2.7 runtime) and one for the latest thing I can find (to be safe if/when App Engine upgrades its runtime).

**Comment by nworden** [comment]
TL;DR for @gimite: is there a partner or anyone else you know of who's using the `charsets` URL parameter?

This is turning out to be a big can of worms, because of the way WebOB 1.2.3 supports non-UTF8 requests. You can't just assign the charset field anymore, you need to manually call `request.decode()`, which doesn't sound so bad except `decode()` doesn't modify the request in place; instead it builds a brand new request and returns that. If your original request object is actually a subclass of the WebOB `BaseRequest` class (say, a webapp `Request` object), any fields from the subclass (which WebOB's `BaseRequest.decode` doesn't know about) get lost. webapp uses at least one of those (the `route` field) for critical functionality.

webapp isn't being maintained, so it's never going to call `Request.decode` in the right place in its own code to deal with it under the hood (which I think is the theoretically correct way this should be handled).

This isn't a long-term problem for us since we're abandoning webapp2 anyway, but it does leave us in a bad position in the meantime. I think there's three options:
1. Do nothing. Hope that App Engine doesn't start using Python 2.7.15 before we've migrated to Django.
2. Abandon support for non-UTF8 requests to make it easier to move to WebOB 1.2.3 (I think we could still support non-UTF8 responses if we wanted).
3. Basically fork webapp2 to address the WebOB change ourselves.

(1) seems dangerous and (3) seems like a lot of work and risk to deal with an edge case with a framework we're already planning to get rid of. Further, it looks like we _currently_ don't support non-UTF8 requests* unless they're specified with the `charsets` param (i.e., setting a non-UTF8 charset with the Content-Type header doesn't seem to work). So, unless there's a partner or someone who knows about and is using our `charsets` URL parameter, there's actually no reduction in functionality from (2).

* I tried running this curl command:
`curl -X GET "http://www.google.org/personfinder/japan/results?query=%8D%B2%93%A1&role=seek" -H "Content-Type: text/html; charset=shift_jis"`
and it shows up in the logs with an error saying "rejecting", from line 249 of `results.py`.

**Comment by gimite** [comment]
I believe it is no longer necessary to support encodings other than UTF-8. But I'm asking @nmakida in a mail thread to confirm because it was originally introduced to support old Japanese feature phones.

FYI Request encoding and response encoding are tightly related because (I believe) browsers send request in the same encoding as the encoding of the page.

**Comment by gimite** [comment]
Confirmed with @nmakida that it should be OK to drop support of non-UTF-8 encodings.

As a reference, this article (found on the web, in Japanese):
https://gist.github.com/uupaa/e19c25029190d2329d8d
says UTF-8 is available for most of  feature phones in Japan sold after 2010.

**Comment by nworden** [comment]
That's great to hear. I'll go ahead with the WebOB upgrade without non-UTF8 support.
I think it could be worth talking about supporting other character sets again once we're on Django though (seems reasonably well-supported, at least from a quick glance at their docs), so I'll leave #579 open for now.
